host/xtest/pkcs11_1000.c

```diff
@@ -8016,3 +8131,102 @@ static void xtest_pkcs11_test_1025(ADBG_Case_t *c)
 
 ADBG_CASE_DEFINE(pkcs11, 1025, xtest_pkcs11_test_1025,
 		 "PKCS11: EDDSA key generation and signing");
+
+static CK_RV test_login_logout_acl_auth(ADBG_Case_t *c, CK_SLOT_ID slot)
+{
+	CK_FLAGS session_flags = CKF_SERIAL_SESSION | CKF_RW_SESSION;
+	CK_SESSION_HANDLE session = CK_INVALID_HANDLE;
+	CK_RV rv = CKR_GENERAL_ERROR;
+
+	Do_ADBG_BeginSubCase(c, "Test C_Login()/C_Logout() with ACL based authentication");
+
+	rv = C_OpenSession(slot, session_flags, NULL, 0, &session);
+	if (!ADBG_EXPECT_CK_OK(c, rv)) goto out;
+
+	/* Logout: should fail as we did not log in yet */
+	rv = C_Logout(session);
+	ADBG_EXPECT_CK_RESULT(c, CKR_USER_NOT_LOGGED_IN, rv);
+
+	/* Login/re-log/logout user */
+	rv = C_Login(session, CKU_USER, NULL, 0);
+	if (!ADBG_EXPECT_CK_OK(c, rv)) goto out_close;
+
+	rv = C_Login(session, CKU_USER, NULL, 0);
+	ADBG_EXPECT_CK_RESULT(c, CKR_USER_ALREADY_LOGGED_IN, rv);
+
+	rv = C_Logout(session);
+	ADBG_EXPECT_CK_OK(c, rv);
+
+out_close:
+	C_CloseSession(session);
+out:
+	Do_ADBG_EndSubCase(c, "Test C_Login()/C_Logout() with ACL based authentication");
+	return rv;
+}
+
+#define GID_STR_LEN  13
+
+static void xtest_pkcs11_test_1026(ADBG_Case_t *c)
+{
+	CK_RV rv = CKR_GENERAL_ERROR;
+	CK_FUNCTION_LIST_PTR ckfunc_list = NULL;
+	CK_SLOT_ID slot = 0;
+	CK_TOKEN_INFO token_info = { };
+	char gid_str[GID_STR_LEN] = { };
+
+	rv = C_GetFunctionList(&ckfunc_list);
+	if (!ADBG_EXPECT_CK_OK(c, rv) ||
+	    !ADBG_EXPECT_NOT_NULL(c, ckfunc_list->C_InitToken) ||
+	    !ADBG_EXPECT_NOT_NULL(c, ckfunc_list->C_InitPIN) ||
+	    !ADBG_EXPECT_NOT_NULL(c, ckfunc_list->C_Login) ||
+	    !ADBG_EXPECT_NOT_NULL(c, ckfunc_list->C_Logout)) {
+		Do_ADBG_Log(" - 1026 -   skip test, ckfunc_list not found");
+		goto out;
```

Comment:
`return rv;` (pkcs11 lib is not yet intialized so we should not call `C_Finalize()` through `close_lib()`).
Ditto for the error case at lines 8187 and 8190 below.

It looks the other `xtest_pkcs11_test_XXXX()` have the same bug. I'll fix taht later.

Reply:
Only "return" might be better without "rv". I'd like to hear your opinion.

Reply:
you're right, without `rv`.